### DISCUSSION TITLE
Avalonia XAML Designer Error "Invalid Markup"  need to implement IActivationForViewFetcher , Autofac/Visual Studio 2022

### DISCUSSION THREAD

**Discussion by agillet76** [AvaloniaUI/Avalonia#7822] [Q&A] [answered]
Hi,
We are developing a desktop app using avalonia, autofac for our inversion control container.
In visual studio 2022, with the latest and previous version of avalonia extensions, the XAML designer always return an Invalid Markup error for any view I implemented:
The error: 
"System.Xaml.XamlException: Don't know how to detect when MyView is activated/deactivated, you may need to implement IActivationForViewFetcher"

We have setup Autofac/Avalonia/ReactiveUI following info from here:
https://stackoverflow.com/questions/65110470/how-to-use-autofac-as-di-container-in-avalonia-reactiveui

Our application works without any issue at runtime, as far as we know.

Anybody encounter the same issue? 
How can we resolve it? We would like to be able to use the designer for our development.

Thanks in advance for any advice

Our view is declared as follow:

public partial class MyView : Avalonia.ReactiveUI.ReactiveUserControl<MyViewModel>

and we use in xaml
<UserControl xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             x:Class="ICS.Devices.UI.Views.MyView"
             mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450">

Here is the Avalonia Diagnostics log:
13:55:25.258 [Verbose]  Started EditorFactory.CreateEditorInstance("C:\Views\MyView.axaml")
13:55:25.258 [Verbose]  Started EditorFactory.GetTextBuffer("C:\Views\MyView.axaml")
13:55:25.258 [Verbose]  Finished EditorFactory.GetTextBuffer("C:\Views\MyView.axaml")
13:55:25.258 [Verbose]  Started EditorFactory.CreateEditorControl()
13:55:25.282 [Verbose]  Finished EditorFactory.CreateEditorControl()
13:55:25.282 [Verbose]  Finished EditorFactory.CreateEditorInstance("C:\Views\MyView.axaml")
13:55:25.283 [Verbose]  Started DesignerPane.Initialize()
13:55:25.294 [Verbose]  Started AvaloniaDesigner.Start()
13:55:25.296 [Verbose]  Started AvaloniaDesigner.LoadTargetsAndStartProcessAsync()
13:55:25.296 [Verbose]  Started AvaloniaDesigner.LoadTargetsAsync()
13:55:25.314 [Verbose]  Finished AvaloniaDesigner.LoadTargetsAsync()
13:55:25.314 [Verbose]  Finished AvaloniaDesigner.LoadTargetsAndStartProcessAsync()
13:55:25.314 [Verbose]  Finished AvaloniaDesigner.Start()
13:55:25.314 [Verbose]  Finished DesignerPane.Initialize()
13:55:25.419 [Verbose]  Started AvaloniaDesigner.StartProcessAsync()
13:55:25.419 [Verbose] 0 Started PreviewerProcess.StartAsync()
13:55:25.420 [Information] 0 Starting previewer process for '"C:\EngineeringApplication\bin\x64\Debug\net6.0\ICS.EngineeringApplication.dll"'
13:55:25.420 [Debug] 0 > dotnet.exe "exec --runtimeconfig \"C:\EngineeringApplication\bin\x64\Debug\net6.0\ICS.EngineeringApplication.runtimeconfig.json\" --depsfile \"C:\EngineeringApplication\bin\x64\Debug\net6.0\ICS.EngineeringApplication.deps.json\" \"C:\Users\Alexandre\.nuget\packages\avalonia\0.10.13\buildTransitive\\..\tools\netcoreapp2.0\designer\Avalonia.Designer.HostApp.dll\" --transport tcp-bson://127.0.0.1:64987/ \"C:\EngineeringApplication\bin\x64\Debug\net6.0\ICS.EngineeringApplication.dll\""
13:55:25.422 [Information] 31872 Started previewer process for '"C:\EngineeringApplication\bin\x64\Debug\net6.0\ICS.EngineeringApplication.dll"'. Waiting for connection to be initialized.
13:55:25.512 [Verbose] 31872 Started PreviewerProcess.ConnectionInitializedAsync()
13:55:25.529 [Debug] 31872 <= "Obtaining AppBuilder instance from ICS.Program.BuildAvaloniaApp"
13:55:25.531 [Information] 31872 Connection initialized
13:55:25.546 [Debug] 31872 => Sending ClientSupportedPixelFormatsMessage { Formats: [Bgra8888, Rgba8888] }
13:55:25.547 [Debug] 31872 => Sending ClientRenderInfoMessage { DpiX: 96, DpiY: 96 }
13:55:25.547 [Verbose] 31872 Finished PreviewerProcess.ConnectionInitializedAsync()
13:55:25.549 [Verbose] 31872 Finished PreviewerProcess.StartAsync()
13:55:25.550 [Debug] 31872 => Sending UpdateXamlMessage { Xaml: "<UserControl xmlns=\"https://git…", AssemblyPath: "C:\Users\Alexandre\source\repos…", XamlFileProjectPath: null }
13:55:25.551 [Verbose]  Finished AvaloniaDesigner.StartProcessAsync()
13:55:25.565 [Debug] 31872 <= "Initializing application in design mode"
13:55:25.940 [Debug] 31872 <= "Sending StartDesignerSessionMessage"
13:55:25.947 [Verbose] 31872 Started PreviewerProcess.OnMessageAsync()
13:55:25.949 [Debug] 31872 <= StartDesignerSessionMessage { SessionId: "a24499f7-1414-4cb6-a0cb-9d09892…" }
13:55:25.950 [Verbose] 31872 Finished PreviewerProcess.OnMessageAsync()
13:55:26.180 [Verbose] 31872 Started PreviewerProcess.OnMessageAsync()
13:55:26.180 [Debug] 31872 <= UpdateXamlResultMessage { Error: "System.Reflection.TargetInvocat…", Handle: null, Exception: ExceptionDetails { ExceptionType: "ArgumentException", Message: "Don't know how to detect when I…", LineNumber: null, LinePosition: null } }
13:55:26.182 [Error] 31872 UpdateXamlResult error
**System.Xaml.XamlException: Don't know how to detect when ICS.Devices.UI.Views.MyView is activated/deactivated, you may need to implement IActivationForViewFetcher**
13:55:26.182 [Verbose] 31872 Finished PreviewerProcess.OnMessageAsync()

**Comment by kekekeks** [comment, marked as answer]
Previewer only calls `BuildAvaloniaApp` and _skips_ your `Main`. Then app initialization goes as usual except no app lifetime is passed to `OnFrameworkInitializationCompleted`. You need to make sure that everything required for your views is initialized during such procedure.

**Reply by agillet76** [reply]
@kekekeks  Thanks for the quick reply and the explanation
I was able to resolve my issue by moving my DI configuration (Autofac) call from Main to inside BuildAvaloniaApp method
